keps/sig-auth/5284-constrained-impersonation/README.md

```diff
@@ -0,0 +1,1445 @@
+
+There are use cases that needs a controller to be able to impersonate:
+1. A controller impersonates a node the controller is running on. This happens when 
+per-node agents (CNI plugins for instance), want to read pods on a given node instead
+of unrestricted pod access.
+2. A controller runs as a deputy, receives request from the user, and impersonates
+the user to start/stop virtual machine or access vm console managed by kubevirt.
+
+This proposal is to introduce additional permissions for impersonation, so that any 
+impersonator can impersonate in a more restricted way.
+
+### Goals
+* Define required permission rules for the impersonator to impersonate 
+  on certain resource/verb.
+* Provide clear example on what permission the impersonator and user should have for the
+  impersonator to impersonate user's certain verbs on certain resources.
+* This is an opt-in feature and existing impersonation flows keep working as-is.
+* Client side semantics of impersonation is unchanged, the existing impersonation headers
+are re-used and no new headers are added.
+
+### Non-Goals
+* Adding impersonator's info to user.Info, so authorization webhooks can know both info
+of impersonated user and the requester.
+
+## Proposal
+
+Introduce a set of verbs with prefix of `impersonate-on:`, e.g. `impersonate-on:create` and
+`impersonate-on:get`. The impersonator needs to have these verbs with certain resources to
+impersonate. 
+
+Introduce verbs `impersonate:user-info`, `impersonate:serviceaccount`， `impersonate:node` and
+`impersonate:scheduled-node`:
+- `impersonate:user-info` limits the impersonator to impersonate users with 
+certain names/groups/userextras. The resources must be `users`/`groups`/`userextras`/`uids`.
+and the user must not be a node (username with a prefix of `system:node:`) and the user must
+not be a service account (username with a prefix of `system:serviceaccount:`)
+The resource names must be usernames, group names or values in the user extras accoringly.
+- `impersonate:serviceaccount` that limits the impersonator to impersonate the serviceaccount with
+the certain name/namespace. The resource must be `serviceaccounts`.
+- `impersonate:node` that limits the impersonator to impersonate the node only. The resource
+must be `nodes`, and the resourceName should be the name of the node. The impersonator must have this
+verb to impersonate a node.
+- `impersonate:scheduled-node` that limits the impersonator to impersonate the node the 
+impersonator is running on. The resources must be `nodes`. For a controller impersonating the node
+that it is running on, it will need to know the node name obtained via downward API:
+```yaml
+env:
+- name: MY_NODE_NAME
+  valueFrom:
+    fieldRef:
+    fieldPath: spec.nodeName
+```
+and then set in the kubeconfig:
+```go
+kubeConfig, _ := clientcmd.BuildConfigFromFlags("", "")
+kubeConfig.Impersonate = rest.ImpersonationConfig{
+	UserName: "system:node:" + os.Getenv("MY_NODE_NAME"),
```

Comment:
Would they also have to add `Groups: ["system:authenticated","system:nodes"]` or not?